From e018ef572ba4ff17caa9e82d5e1b5cea0d76f903 Mon Sep 17 00:00:00 2001
From: kevin Barnett <kevin.barnett@microsemi.com>
Date: Fri, 16 Sep 2016 15:01:51 -0500
Subject: [PATCH 16/16] scsi: smartpqi: raid bypass lba calculation fix

In the ioaccel path, the calculation of the starting LBA for
READ(6)/WRITE(6) SCSI commands does not take into account the most
significant 5 bits of the LBA: it only uses the least significant 16
bits of the starting LBA.

Reported-by: Mahesh Rajashekhara <mahesh.rajashekhara@microsemi.com>
Signed-off-by: Kevin Barnett <kevin.barnett@microsemi.com>
Signed-off-by: Don Brace <don.brace@microsemi.com>
Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
---
 drivers/scsi/smartpqi/smartpqi_init.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/scsi/smartpqi/smartpqi_init.c b/drivers/scsi/smartpqi/smartpqi_init.c
index 52cfa26..a535b26 100644
--- a/drivers/scsi/smartpqi/smartpqi_init.c
+++ b/drivers/scsi/smartpqi/smartpqi_init.c
@@ -1954,7 +1954,8 @@ static int pqi_raid_bypass_submit_scsi_cmd(struct pqi_ctrl_info *ctrl_info,
 		is_write = true;
 		/* fall through */
 	case READ_6:
-		first_block = (u64)get_unaligned_be16(&scmd->cmnd[2]);
+		first_block = (u64)(((scmd->cmnd[1] & 0x1f) << 16) |
+			(scmd->cmnd[2] << 8) | scmd->cmnd[3]);
 		block_cnt = (u32)scmd->cmnd[4];
 		if (block_cnt == 0)
 			block_cnt = 256;
-- 
1.7.10.4

